sklearn_numba_dpex/kmeans/drivers.py

```diff
@@ -718,26 +745,35 @@ def _set_dtype(self, X, centers_init):
                 f"The centers have been passed with type {centers_init_dtype} but "
                 f"type {compute_dtype} is expected. A copy will be created with the "
                 f"correct type {compute_dtype}. Ensure that the centers are passed "
-                f"with the correct dtype to save memory and disable this warning.",
+                f"with the correct dtype to save memory and suppress this warning.",
                 DataConversionWarning,
             )
             centers_init = centers_init.astype(compute_dtype)
 
-        return X, centers_init, compute_dtype, output_dtype
-
-    def _check_inputs(self, X, cluster_centers, sample_weight):
-        if (sample_weight is not None) and ((sample_weight != sample_weight[0]).any()):
-            raise NotSupportedByEngineError(
-                "The sklearn_numba_dpex engine for KMeans does not support "
-                "non-uniform sample weights at this time, but it is a planned "
-                "feature. More information about the progress toward support is "
-                "available on the issue tracker at: "
-                "https://github.com/soda-inria/sklearn-numba-dpex/issues/22  ."
+        if sample_weight is not None and sample_weight.dtype != compute_dtype:
+            warnings.warn(
+                f"sample_weight has been passed with type {sample_weight.dtype} but "
+                f"type {compute_dtype} is expected. A copy will be created with the "
+                f"correct type {compute_dtype}. Ensure that sample_weight is passed "
+                f"with the correct dtype to save memory and suppress this warning.",
+                DataConversionWarning,
             )
+            sample_weight = sample_weight.astype(compute_dtype)
 
-        X, cluster_centers, compute_dtype, output_dtype = self._set_dtype(
-            X, cluster_centers
-        )
+        return X, sample_weight, centers_init, compute_dtype, output_dtype
+
+    def _check_inputs(self, X, sample_weight, cluster_centers):
+
+        if sample_weight is not None and not (sample_weight != sample_weight[0]).any():
+            sample_weight = None
```

Comment:
If `sample_weight` is provided as a vector of equal weights, I think we still need to treat it as such because this might be an (I agree weird) intention from the user. In this case, I would change this test to only perform this optimisation in case of unary weights:
```suggestion
        # If sample weights are unary, we can simply ignore them.
        if sample_weight is not None and not (sample_weight != 1).any():
            sample_weight = None
```

Reply:
Actually I think we should not set `sample_weight = None` even when `sample_weight == np.ones(n_samples)`.

This will make it possible to check that the computation with `sample_weight is None` returns the same centroids / labels as with `sample_weight == np.ones(n_samples)` while being faster.

Reply:
middle ground: still apply the optimization but multiply the total inertia by the constant just before it's printed / returned (from the user point of view, will have the same effect)

Reply:
> Actually I think we should not set `sample_weight = None` even when `sample_weight == np.ones(n_samples)`.

@ogrisel the issue is that sklearn always transforms a `None` into an array of ones, if we don't force this optimization here, with the current code it will never happen because sklearn force the opposite (and our engine currently inherits from all this preparation)

Reply:
Ok then. Have you measured the performance impact? If it's not significant, we might was well do as in the original scikit-learn code base and drop the unweighted kernels.